### PR TITLE
Use `PropertyResourceBundle` in test suite and validate `.properties` files are not both valid UTF-8 and valid ISO-8859-1

### PR DESCRIPTION
**Pull request by timja**
see https://github.com/jenkinsci/jenkins/pull/6550#issuecomment-1121055141 https://github.com/jenkinsci/stapler/pull/357

You can't actually use a file that is valid utf-8 and not valid ISO-8859-1 in at least core right now due to https://github.com/jenkinsci/jenkins/blob/ff39fbb3cdf58d7dda17c37b8d0606dc8e9282f4/core/pom.xml#L604

But we can adjust that later when we switch to Java 11

Validated with core:
```
[ERROR] Tests run: 7139, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.743 s <<< FAILURE! - in jenkins.CorePropertiesTest
[ERROR] hudson/Messages_es.properties(org.jvnet.hudson.test.PropertiesTestSuite$PropertiesTest)  Time elapsed: 0.004 s  <<< FAILURE!
java.lang.AssertionError: jar:file:/Users/timja/.m2/repository/org/jenkins-ci/main/jenkins-core/2.347-SNAPSHOT/jenkins-core-2.347-SNAPSHOT.jar!/hudson/Messages_es.properties is encoded with both valid utf-8 and ISO-8859-1 characters, the utf-8 characters need to be encoded with something like `native2ascii`
	at org.jvnet.hudson.test.PropertiesTestSuite$PropertiesTest.runTest(PropertiesTestSuite.java:87)
```

### CODE REVIEW

File: src/main/java/org/jvnet/hudson/test/PropertiesTestSuite.java [hunk 1]
```diff
@@ -66,11 +78,43 @@ public synchronized Object put(Object key, Object value) {
                     return null;
                 }
             };
+
+            byte[] contents = IOUtils.toByteArray(resource.openStream());
+            if (!isEncoded(contents, StandardCharsets.US_ASCII)) {
+                boolean isUtf8 = isEncoded(contents, StandardCharsets.UTF_8);
+                boolean isIso88591 = isEncoded(contents, StandardCharsets.ISO_8859_1);
+                if (isUtf8 && isIso88591) {
+                    throw new AssertionError(resource + " is encoded with both valid utf-8 and ISO-8859-1 characters, the utf-8 characters need to be encoded with something like `native2ascii`");
```
Comment (basil) [inline src/main/java/org/jvnet/hudson/test/PropertiesTestSuite.java:87]: The message seems a bit confusing to me as there is really no such thing as a "UTF-8 character" or a "ISO-8859-1" character. In Java all strings are Unicode. I would rephrase this as follows (suggestion only, I know this is a bit verbose):

```suggestion
                    throw new AssertionError(resource + " is valid UTF-8 and valid ISO-8859-1. To avoid problems when auto-detecting the encoding, use the lowest common denominator of ASCII encoding and express non-ASCII characters with escape sequences using a tool like `native2ascii`.");
```

Feel free to change this if you want to make it shorter.

**Review by basil — APPROVED** [review]
Thanks for the PR!

File: src/main/java/org/jvnet/hudson/test/PropertiesTestSuite.java [hunk 2]
```diff
@@ -66,11 +78,43 @@ public synchronized Object put(Object key, Object value) {
                     return null;
                 }
             };
+
+            byte[] contents = IOUtils.toByteArray(resource.openStream());
```
[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/org/jvnet/hudson/test/PropertiesTestSuite.java:82]:
```suggestion
            byte[] contents = IOUtils.toByteArray(resource);
```

suffices.

**Review by jglick — APPROVED** [review]
Other than the above nits, looks nice! Thanks for taking care of this.

File: src/main/java/org/jvnet/hudson/test/PropertiesTestSuite.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (basil) [reply src/main/java/org/jvnet/hudson/test/PropertiesTestSuite.java:82]: Even better.

**Review by jglick — APPROVED** [review]

File: src/main/java/org/jvnet/hudson/test/PropertiesTestSuite.java [hunk 3]
```diff
@@ -66,11 +78,43 @@ public synchronized Object put(Object key, Object value) {
...
-                props.load(is);
+                PropertyResourceBundle propertyResourceBundle = new PropertyResourceBundle(is);
+                Enumeration<String> keys = propertyResourceBundle.getKeys();
+                // TODO Java 9+ can use 'asIterator' and get rid of below collections conversion
+                List<String> keysAsSaneType = Collections.list(keys);
+
+                for (String localKey : keysAsSaneType) {
+                    String value = propertyResourceBundle.getString(localKey);
+                    props.setProperty(localKey, value);
+                }
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/PropertiesTestSuite.java]: https://github.com/jenkinsci/jenkins-test-harness/issues/1141
